core/corehttp/gateway_test.go

```diff
@@ -20,18 +21,33 @@ import (
 	testutil "github.com/ipfs/go-ipfs/util/testutil"
 )
 
-type mockNamesys map[string]path.Path
+type mockNamesys map[string]mockEntry
 
-func (m mockNamesys) Resolve(ctx context.Context, name string) (value path.Path, err error) {
-	return m.ResolveN(ctx, name, namesys.DefaultDepthLimit)
+type mockEntry struct {
+	path path.Path
+	ttl  time.Duration
 }
 
-func (m mockNamesys) ResolveN(ctx context.Context, name string, depth int) (value path.Path, err error) {
-	p, ok := m[name]
+func (m mockNamesys) Resolve(ctx context.Context, name string) (path.Path, error) {
+	res, err := m.ResolveDetailed(ctx, name)
+	return res.Path, err
+}
+
+func (m mockNamesys) ResolveN(ctx context.Context, name string, depth int) (path.Path, error) {
+	res, err := m.ResolveNDetailed(ctx, name, depth)
+	return res.Path, err
+}
+
+func (m mockNamesys) ResolveDetailed(ctx context.Context, name string) (namesys.ResolveResult, error) {
+	return m.ResolveNDetailed(ctx, name, namesys.DefaultDepthLimit)
+}
+
+func (m mockNamesys) ResolveNDetailed(ctx context.Context, name string, depth int) (namesys.ResolveResult, error) {
```

Comment:
The TTL field is actually useful in error cases. Many errors should not be cached (such as when we fail to fetch something from the network but it could be available a second later) but some errors could certainly be cached:

* `/ipfs/QmFoo/file` is requested; `/ipfs/QmFoo` resolves correctly (TTL=“infinite”) but does not contain `file`: the error’s TTL can be “infinite”.
* `/ipns/QmBar/file` is requested; `/ipns/QmBar` resolves to `/ipfs/QmFoo` (TTL=2 minutes for example) but that does not contain `file`: the error’s TTL=2 minute because that’s how long the resolved IPNS entry can be cached.
* A parse error in the path being resolved: the error’s TTL can be “infinite”. e.g. `namesys/namesys.go`:

```go
        segments := strings.SplitN(name, "/", 3)
        if len(segments) < 3 || segments[0] != "" {
                log.Warningf("Invalid name syntax for %s", name)
                return ResolveResult{TTL: ImmutableTTL}, ErrResolveFailed
```

Nothing in the code caches errors or serves HTTP errors with a `max-age` *yet* but the infrastructure is there if we choose to implement caching for some errors later.

Would it be a reasonable solution to return `ttl time.Duration, result *ResolveResult, err error` where `ttl` is always given (usually 0 for errors but not always) and exactly one of `result`, `err` is nil?